### PR TITLE
Build examples from documentation in CI

### PR DESCRIPTION
**Pull request by lieser**
Extract the examples in the documentation into separate files and add a standalone Makefile for them.
This allows us to build them in the CI and gives us at least a minimal insurance that they actually work.

A limitation of the current approach is that all examples have to be standalone executable with a main.

TODOs:

- [x] Add new code structure to documentation.
- [x] Extract examples in `doc/api_ref/pkcs11.rst`.
    - Examples would need some changes, as the current ones to not have any includes or main.

Open Questions:

- How to properly model the requirements of the examples? Currently the examples target would simply fail if Botan is not configured with all required modules and boost enabled.
    - Solution 1:
        - Do the same as in tests with using `#if defined(BOTAN_HAS_NUMBERTHEORY)` and similar things
        - We could use [`start-after` / `end-before`](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-literalinclude) in the `literalinclude` directive to avoid it being shown in the documentation
        - Would still look ugly if looking directly at the source file
        - We would create "empty" example executable
    - Are there better but still simple solutions?
    - Is this even a concern for us? E.g. for the CI this could be a desired behavior, as no example is silently excluded from it
    - **Decision:** This should no longer be a concern as we removed the special target that exposed the examples
- Also run examples? Would of course be nice, but I would leave that for later.

### CODE REVIEW

**Comment by reneme** [comment]
> Are there better but still simple solutions?

I'm not sure about promoting those example-builds as a full-blown build target for two reasons:

1. Users and distributers would need to be educated that this build target is somewhat internal and will not result in any meaningful build artefact
2. The "examples" build should as much as possible mimick an actual library user. Which might not be the case when they are built via Botan's main build system.

Instead, let's place a simple Makefile next to the example source files and invoke that exclusively in a CI job from `ci_build.py`. The CI can then make sure that the library is configured sufficiently and we avoid the confusion of the new first-class build target.

**Comment by randombit** [comment]
This is great! Being able to test documentation examples is a feature I really like about `rustdoc` and have missed in C++. Also agree with @reneme that this would be confusing as a toplevel target that is visible to end users - building these examples for test purposes should only ever happen in our CI.

**Comment by lieser** [comment]
Replaced the added target with a separate Makefile in the examples directory.

**Comment by lieser** [comment]
I think this is now ready for a review.
Note that I tried to keep the code move in separate commits. So looking at the commits separately should make it easier to see what I changed in the examples.

File: src/examples/tls_custom_curves_server.cpp [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+         }
+
+      std::vector<Botan::Certificate_Store*> trusted_certificate_authorities(
+         const std::string& /*type*/,
+         const std::string& /*context*/) override
+         {
+         // if client authentication is required, this function
+         // shall return a list of certificates of CAs we trust
+         // for tls client certificates, otherwise return an empty list
+         return std::vector<Botan::Certificate_Store*>();
```
Comment (reneme) [inline src/examples/tls_custom_curves_server.cpp:89]:
```suggestion
         return {};
```

File: src/examples/tls_proxy.cpp [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+         }
+
+      std::vector<Botan::Certificate_Store*> trusted_certificate_authorities(
+         const std::string& /*type*/,
+         const std::string& /*context*/) override
+         {
+         // if client authentication is required, this function
+         // shall return a list of certificates of CAs we trust
+         // for tls client certificates, otherwise return an empty list
+         return std::vector<Botan::Certificate_Store*>();
```
Comment (reneme) [inline src/examples/tls_proxy.cpp:72]:
```suggestion
         return {};
```

File: src/examples/tls_proxy.cpp [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+                           policy,
+                           rng);
+
+   // read data received from the tls client, e.g., using BSD sockets or boost asio
+   // and pass it to server.received_data().
+   // ...
+
+   // send data to the tls client using server.send()
+   // ...
+   }
```
Comment (reneme) [inline src/examples/tls_proxy.cpp:120]:
```suggestion
   
   return 0;
   }
```

File: src/examples/tls_stream_client.cpp [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+   http::request<http::string_body> req;
+   req.version(11);
+   req.method(http::verb::get);
+   req.target("/news.html");
+   req.set(http::field::host, "botan.randombit.net");
+
+   client c(io_context, iterator, req);
+
+   io_context.run();
+   }
```
Comment (reneme) [inline src/examples/tls_stream_client.cpp:126]:
```suggestion
   
   return 0;
   }
```

File: src/examples/xmss.cpp [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+   // create Public Key Verifier using the public key
+   Botan::PK_Verifier verifier(public_key, "");
+
+   // verify the signature for the previously generated message.
+   verifier.update(msg.data(), msg.size());
+   if(verifier.check_signature(sig.data(), sig.size()))
+      {
+      std::cout << "Success." << std::endl;
+      }
```
Comment (reneme) [inline src/examples/xmss.cpp:37]:
```suggestion
      return 0;
      }
```

File: src/examples/xmss.cpp [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+   else
+      {
+      std::cout << "Error." << std::endl;
+      }
```
Comment (reneme) [inline src/examples/xmss.cpp:41]:
```suggestion
      return 1;
      }
```

File: src/examples/Makefile [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+# Paths to relevant programs
+CXX            = g++
+LINKER         = $(CXX)
+
+# Compiler Flags
+ABI_FLAGS      = -fstack-protector -m64 -pthread
+LANG_FLAGS     = -std=c++17 -D_REENTRANT
+LANG_EXE_FLAGS =
+CXXFLAGS       = -O3
+WARN_FLAGS     = -Wall -Wextra -Wpedantic -Wstrict-aliasing -Wcast-align -Wmissing-declarations -Wpointer-arith -Wcast-qual -Wzero-as-null-pointer-constant -Wnon-virtual-dtor -Wold-style-cast -Wsuggest-override -Wshadow -Wextra-semi -Werror -Wno-error=strict-overflow -Wno-error=zero-as-null-pointer-constant -Wno-error=non-virtual-dtor
```
Comment (reneme) [inline src/examples/Makefile:15]: For the sake of readability: Please add `-Wno-unused-parameter` and remove the commenting of unused functions parameters in the examples.

**Review by reneme — COMMENTED** [review]
Thanks a lot for setting this up!

I used the opportunity to comb through the example code and left a bunch of remarks. Though, it would be totally fair to outsource this into an extra PR.

Also, the coding style of the examples is really inconsistent. I'd suggest to run `clang-format` on it with a popular setting (maybe "Google" or "LLVM" for it's 2-space indents).

Obviously, it would now be great to also run the examples (and expect a zero status code), but that's out of scope of this PR.

File: src/examples/check_key.cpp [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+int main()
+   {
+   Botan::X509_Certificate cert("cert.pem");
+   std::unique_ptr<Botan::RandomNumberGenerator> rng(new Botan::AutoSeeded_RNG);
+   std::unique_ptr<Botan::Public_Key> key(cert.subject_public_key());
+   if(!key->check_key(*rng.get(), false))
+      {
+      throw std::invalid_argument("Loaded key is invalid");
+      }
+   }
```
[the remark this answers is not in the document]
Comment (lieser) [reply src/examples/check_key.cpp:17]: changed

File: src/examples/hmac.cpp [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+
+   // "Message" != "Mussage" so tags will also not match
+   std::string tag1 = compute_mac("Message", key);
+   std::string tag2 = compute_mac("Mussage", key);
+   assert(tag1 != tag2);
+
+   // Recomputing with original input message results in identical tag
+   std::string tag3 = compute_mac("Message", key);
+   assert(tag1 == tag3);
+   }
```
[the remark this answers is not in the document]
Comment (lieser) [reply src/examples/hmac.cpp:33]: changed

File: src/examples/pkcs11_low_level.cpp [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+
+   // C_Initialize is automatically called by the constructor of the Module
+
+   // work with the token
+
+   std::vector<Botan::PKCS11::SlotId> slot_ids;
+   [[maybe_unused]] bool success = module->C_GetSlotList(true, slot_ids);
+
+   // C_Finalize is automatically called by the destructor of the Module
+   }
```
[the remark this answers is not in the document]
Comment (lieser) [reply src/examples/pkcs11_low_level.cpp:20]: changed

File: src/examples/pkcs11_module.cpp [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+
+   // Sometimes useful if a newly connected token is not detected by the PKCS#11 module
+   module.reload();
+
+   Botan::PKCS11::Info info = module.get_info();
+
+   // print library version
+   std::cout << std::to_string( info.libraryVersion.major ) << "."
+      << std::to_string( info.libraryVersion.minor ) << std::endl;
+   }
```
[the remark this answers is not in the document]
Comment (lieser) [reply src/examples/pkcs11_module.cpp:21]: changed

File: src/examples/pkcs11_objects.cpp [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+
+   // search for an object
+   Botan::PKCS11::AttributeContainer search_template;
+   search_template.add_string( Botan::PKCS11::AttributeType::Label, "Botan" );
+   auto found_objs =
+      Botan::PKCS11::Object::search<Botan::PKCS11::Object>( session, search_template.attributes() );
+
+   // destroy the object
+   data_obj.destroy();
+   }
```
[the remark this answers is not in the document]
Comment (lieser) [reply src/examples/pkcs11_objects.cpp:61]: changed

File: src/examples/pkcs11_rng.cpp [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+
+   /************ add entropy *************/
+   Botan::AutoSeeded_RNG auto_rng;
+   auto auto_rng_random = auto_rng.random_vec( 20 );
+   p11_rng.add_entropy( auto_rng_random.data(), auto_rng_random.size() );
+
+   /************ use PKCS#11 RNG to seed HMAC_DRBG *************/
+   Botan::HMAC_DRBG drbg( Botan::MessageAuthenticationCode::create( "HMAC(SHA-512)" ), p11_rng );
+   drbg.randomize( random.data(), random.size() );
+   }
```
[the remark this answers is not in the document]
Comment (lieser) [reply src/examples/pkcs11_rng.cpp:34]: changed

File: src/examples/pkcs11_session.cpp [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+   // logoff
+   session.logoff();
+
+   // log in as security officer
+   Botan::PKCS11::secure_string so_pin = { '0', '0', '0', '0', '0', '0', '0', '0' };
+   session.login( Botan::PKCS11::UserType::SO, so_pin );
+
+   // change pin to old pin
+   session.init_pin( pin );
+   }
```
[the remark this answers is not in the document]
Comment (lieser) [reply src/examples/pkcs11_session.cpp:65]: changed

File: src/examples/pkcs11_slot.cpp [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+   Botan::PKCS11::MechanismInfo mech_info =
+      slot.get_mechanism_info( Botan::PKCS11::MechanismType::RsaPkcsOaep );
+
+   // maximum RSA key length supported:
+   std::cout << mech_info.ulMaxKeySize << std::endl;
+
+   // initialize the token
+   Botan::PKCS11::secure_string so_pin( 8, '0' );
+   slot.initialize( "Botan PKCS11 documentation test label", so_pin );
+   }
```
[the remark this answers is not in the document]
Comment (lieser) [reply src/examples/pkcs11_slot.cpp:43]: changed

File: src/examples/pkcs11_token_management.cpp [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+
+   // change back
+   Botan::PKCS11::change_pin( slot, test_pin, pin );
+
+   /************ change security officer pin *************/
+   Botan::PKCS11::change_so_pin( slot, so_pin, test_pin );
+
+   // change back
+   Botan::PKCS11::change_so_pin( slot, test_pin, so_pin );
+   }
```
[the remark this answers is not in the document]
Comment (lieser) [reply src/examples/pkcs11_token_management.cpp:44]: changed

File: src/examples/pkcs11_x509.cpp [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+   props.set_label( "Botan PKCS#11 test certificate" );
+   props.set_private( false );
+   props.set_token( true );
+
+   // import
+   Botan::PKCS11::PKCS11_X509_Certificate pkcs11_cert( session, props );
+
+   // load by handle
+   Botan::PKCS11::PKCS11_X509_Certificate pkcs11_cert2( session, pkcs11_cert.handle() );
+   }
```
[the remark this answers is not in the document]
Comment (lieser) [reply src/examples/pkcs11_x509.cpp:35]: changed

File: src/examples/tls_client.cpp [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+   Botan::TLS::Strict_Policy policy;
+
+   // open the tls connection
+   Botan::TLS::Client client(callbacks,
+                           session_mgr,
+                           creds,
+                           policy,
+                           rng,
+                           Botan::TLS::Server_Information("botan.randombit.net", 443),
+                           Botan::TLS::Protocol_Version::TLS_V12);
```
[the remark this answers is not in the document]
Comment (lieser) [reply src/examples/tls_client.cpp:103]: Decided to leave it as TLS 1.2 for now.

File: src/examples/tls_custom_curves_client.cpp [hunk 19]
```diff
@@ -0,0 +1,168 @@
...
+
+   while(!client.is_closed())
+      {
+      // read data received from the tls server, e.g., using BSD sockets or boost asio
+      // ...
+
+      // send data to the tls server using client.send()
+
+      }
+   }
```
[the remark this answers is not in the document]
Comment (lieser) [reply src/examples/tls_custom_curves_client.cpp:173]: changed

File: src/examples/tls_custom_curves_server.cpp [hunk 20]
```diff
@@ -0,0 +1,179 @@
...
+                           policy,
+                           rng);
+
+   // read data received from the tls client, e.g., using BSD sockets or boost asio
+   // and pass it to server.received_data().
+   // ...
+
+   // send data to the tls client using server.send()
+   // ...
+   }
```
[the remark this answers is not in the document]
Comment (lieser) [reply src/examples/tls_custom_curves_server.cpp:181]: changed

File: src/examples/dl_group.cpp [hunk 21]
```diff
@@ -0,0 +1,17 @@
+#include <botan/dl_group.h>
+#include <botan/auto_rng.h>
+#include <botan/rng.h>
+
+#include <iostream>
+
+int main()
+   {
+      std::unique_ptr<Botan::RandomNumberGenerator> rng(new Botan::AutoSeeded_RNG);
+      std::unique_ptr<Botan::DL_Group> group(new Botan::DL_Group(*rng.get(), Botan::DL_Group::Strong, 2048));
```
[the remark this answers is not in the document]
Comment (lieser) [reply src/examples/dl_group.cpp]: changed

File: src/examples/check_key.cpp [hunk 22]
```diff
@@ -0,0 +1,15 @@
...
+#include <botan/auto_rng.h>
+#include <botan/rng.h>
+#include <botan/pk_keys.h>
+
+int main()
+   {
+   Botan::X509_Certificate cert("cert.pem");
+   std::unique_ptr<Botan::RandomNumberGenerator> rng(new Botan::AutoSeeded_RNG);
+   std::unique_ptr<Botan::Public_Key> key(cert.subject_public_key());
+   if(!key->check_key(*rng.get(), false))
```
[the remark this answers is not in the document]
Comment (lieser) [reply src/examples/check_key.cpp]: changed to allocation on stack

File: src/examples/encrypt_with_pkcs8_key.cpp [hunk 23]
```diff
@@ -0,0 +1,31 @@
...
+   //load keypair
+   Botan::DataSource_Stream in(argv[1]);
+   std::unique_ptr<Botan::Private_Key> kp(Botan::PKCS8::load_key(in));
+
+   //encrypt with pk
+   Botan::PK_Encryptor_EME enc(*kp,*rng.get(), "EME1(SHA-256)");
+   std::vector<uint8_t> ct = enc.encrypt(pt,*rng.get());
+
+   //decrypt with sk
+   Botan::PK_Decryptor_EME dec(*kp,*rng.get(), "EME1(SHA-256)");
```
[the remark this answers is not in the document]
Comment (lieser) [reply src/examples/encrypt_with_pkcs8_key.cpp]: changed

File: src/examples/encrypt_with_pkcs8_key.cpp [hunk 24]
```diff
@@ -0,0 +1,31 @@
...
+   std::cout << std::endl << "enc: " << Botan::hex_encode(ct) << std::endl << "dec: "<< Botan::hex_encode(dec.decrypt(ct));
```
[the remark this answers is not in the document]
Comment (lieser) [reply src/examples/encrypt_with_pkcs8_key.cpp]: changed

File: src/examples/hmac.cpp [hunk 25]
```diff
@@ -0,0 +1,31 @@
...
+
+   hmac->set_key(key);
+   hmac->update(msg);
+
+   return Botan::hex_encode(hmac->final());
+   }
+
+int main()
+   {
+   Botan::System_RNG rng;
```
[the remark this answers is not in the document]
Comment (lieser) [reply src/examples/hmac.cpp]: changed

File: src/examples/pkcs11_ecdh.cpp [hunk 26]
```diff
@@ -0,0 +1,119 @@
...
+
+   Botan::SymmetricKey alice_key = ka.derive_key( 32,
+      key_pair_other.first.public_point().encode(Botan::PointGFp::UNCOMPRESSED));
+
+   Botan::SymmetricKey bob_key = kb.derive_key( 32,
+      key_pair.first.public_point().encode(Botan::PointGFp::UNCOMPRESSED));
+
+   bool eq = alice_key == bob_key;
+
+   return eq;
```
[the remark this answers is not in the document]
Comment (lieser) [reply src/examples/pkcs11_ecdh.cpp]: fixed

File: src/examples/pkcs11_ecdsa.cpp [hunk 27]
```diff
@@ -0,0 +1,110 @@
...
+
+   std::vector<uint8_t> plaintext( 20, 0x01 );
+
+   Botan::PK_Signer signer( key_pair.second, rng, "Raw", Botan::IEEE_1363, "pkcs11" );
+   auto signature = signer.sign_message( plaintext, rng );
+
+   Botan::PK_Verifier token_verifier( key_pair.first, "Raw", Botan::IEEE_1363, "pkcs11" );
+   bool ecdsa_ok = token_verifier.verify_message( plaintext, signature );
+
+   return ecdsa_ok;
```
[the remark this answers is not in the document]
Comment (lieser) [reply src/examples/pkcs11_ecdsa.cpp]: fixed

File: src/examples/pkcs11_rng.cpp [hunk 28]
```diff
@@ -0,0 +1,33 @@
+
```
[the remark this answers is not in the document]
Comment (lieser) [reply src/examples/pkcs11_rng.cpp]: fixed

File: src/examples/pkcs11_rsa.cpp [hunk 29]
```diff
@@ -0,0 +1,106 @@
...
+
+   Botan::PK_Signer signer( rsa_keypair.second, rng, "EMSA4(SHA-256)", Botan::IEEE_1363 );
+   auto signature = signer.sign_message( plaintext, rng );
+
+   /************ RSA verify *************/
+
+   Botan::PK_Verifier verifier( rsa_keypair.first, "EMSA4(SHA-256)", Botan::IEEE_1363 );
+   auto ok = verifier.verify_message( plaintext, signature );
+
+   return ok;
```
[the remark this answers is not in the document]
Comment (lieser) [reply src/examples/pkcs11_rsa.cpp]: fixed

File: src/examples/tls_client.cpp [hunk 30]
```diff
@@ -0,0 +1,117 @@
...
+ * It returns a list of trusted CA certificates from a local directory.
+ * TLS client authentication is disabled. See src/lib/tls/credentials_manager.h.
+ */
+class Client_Credentials : public Botan::Credentials_Manager
+   {
+   public:
+      Client_Credentials()
+         {
+         // Here we base trust on the system managed trusted CA list
+         m_stores.push_back(new Botan::System_Certificate_Store);
```
[the remark this answers is not in the document]
Comment (lieser) [reply src/examples/tls_client.cpp]: Agree that memory leaks in examples are bad. Should be fixed now. I removed all occurrences of *new* from the examples.

File: src/examples/tls_client.cpp [hunk 31]
```diff
@@ -0,0 +1,117 @@
...
+
+      std::vector<Botan::X509_Certificate> cert_chain(
+         const std::vector<std::string>& /*cert_key_types*/,
+         const std::string& /*type*/,
+         const std::string& /*context*/) override
+         {
+         // when using tls client authentication (optional), return
+         // a certificate chain being sent to the tls server,
+         // else an empty list
+         return std::vector<Botan::X509_Certificate>();
```
[the remark this answers is not in the document]
Comment (lieser) [reply src/examples/tls_client.cpp]: changed

File: src/examples/tls_custom_curves_client.cpp [hunk 32]
```diff
@@ -0,0 +1,168 @@
...
+class Client_Credentials : public Botan::Credentials_Manager
+{
+   public:
+      std::vector<Botan::Certificate_Store*> trusted_certificate_authorities(
+         const std::string& /*type*/,
+         const std::string& /*context*/) override
+         {
+         // return a list of certificates of CAs we trust for tls server certificates,
+         // e.g., all the certificates in the local directory "cas"
+         return { new Botan::Certificate_Store_In_Memory("cas") };
```
[the remark this answers is not in the document]
Comment (lieser) [reply src/examples/tls_custom_curves_client.cpp]: changed

File: src/examples/tls_custom_curves_client.cpp [hunk 33]
```diff
@@ -0,0 +1,168 @@
...
+
+      std::vector<Botan::X509_Certificate> cert_chain(
+         const std::vector<std::string>& /*cert_key_types*/,
+         const std::string& /*type*/,
+         const std::string& /*context*/) override
+         {
+         // when using tls client authentication (optional), return
+         // a certificate chain being sent to the tls server,
+         // else an empty list
+         return std::vector<Botan::X509_Certificate>();
```
[the remark this answers is not in the document]
Comment (lieser) [reply src/examples/tls_custom_curves_client.cpp]: changed

File: src/examples/tls_custom_curves_server.cpp (hunk 1 above)
Comment (lieser) [reply src/examples/tls_custom_curves_server.cpp]: changed

File: src/examples/tls_proxy.cpp (hunk 2 above)
Comment (lieser) [reply src/examples/tls_proxy.cpp]: changed

File: src/examples/tls_stream_client.cpp [hunk 34]
```diff
@@ -0,0 +1,126 @@
...
+      http::response<http::string_body> response_;
+      boost::beast::flat_buffer         reply_;
+
+      Botan::TLS::Session_Manager_Noop session_mgr_;
+      Botan::AutoSeeded_RNG            rng_;
+      Credentials_Manager              credentials_mgr_;
+      Botan::TLS::Policy               policy_;
+
+      Botan::TLS::Context                              ctx_;
+      Botan::TLS::Stream<boost::asio::ip::tcp::socket> stream_;
```
[the remark this answers is not in the document]
Comment (lieser) [reply src/examples/tls_stream_client.cpp]: changed

File: src/examples/Makefile (hunk 7 above)
Comment (lieser) [reply src/examples/Makefile]: changed

File: src/examples/tls_client.cpp (hunk 30 above)
[the remark this answers is not in the document]
Comment (reneme) [reply src/examples/tls_client.cpp]: Thanks!

**Review by reneme — APPROVED** [review]

**Pull request merged**